lib/block_view.go

```diff
@@ -2987,24 +3067,74 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
 
-func _verifySignature(txn *MsgBitCloutTxn) error {
-	// Compute a hash of the transaction
+func (bav *UtxoView) _verifySignature(txn *MsgBitCloutTxn, blockHeight uint32) error {
+	// Compute a hash of the transaction.
 	txBytes, err := txn.ToBytes(true /*preSignature*/)
 	if err != nil {
 		return errors.Wrapf(err, "_verifySignature: Problem serializing txn without signature: ")
 	}
 	txHash := Sha256DoubleHash(txBytes)
-	// Convert the txn public key into a *btcec.PublicKey
-	txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+
+	// We make a local copy of the signature as it should be read-only.
+	sigCopy := append([]byte{}, txn.Signature...)
+	// Verify that the control byte is within the allowed range.
+	// DERControlByte means txn was signed with the owner's private key.
+	// Otherwise, txn was signed with a derived key and we will attempt
+	// to recover the public key from the signature.
+	sigByte := sigCopy[0]
+	if !(sigByte >= DERControlByte && sigByte <= SIGMaxByte) {
+		return fmt.Errorf("_verifySignature: Signature control byte outside of allowed range")
+	}
+
+	// Reset the control byte to the DER control byte. Verify signature encoding.
+	sigCopy[0] = DERControlByte
+	sig, err := btcec.ParseDERSignature(sigCopy, btcec.S256())
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing signature: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// We check for public key solution iteration for this transaction. This
+	// allows us to recover public key from the signature.
+	// If the transaction was signed with the owner key, iteration is equal to 0.
+	// If the transaction was signed with a derived key, iteration is between 1-4.
+	iteration := sigByte - DERControlByte
+	if iteration == 0 {
+		// Convert the txn public key into a *btcec.PublicKey
+		txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+		if err != nil {
+			return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		}
+		// Verify that the transaction is signed by the specified key.
+		if sig.Verify(txHash[:], txnPk) {
+			return nil
+		}
+	} else {
+		// Subtract 1 from iteration as it should be between 0-3.
+		iteration = iteration - 1
+
+		// Recover public key from the signature
+		sigCompact := SignatureSerializeCompactWithIteration(sig, int(iteration), false)
+		derivedPk, _, err := btcec.RecoverCompact(btcec.S256(), sigCompact, txHash[:])
+		if err != nil {
+			return errors.Wrapf(err, "_verifySignature: Problem recovering public key: ")
+		}
+
+		ownerPublicKey := txn.PublicKey
+		derivedPublicKey := derivedPk.SerializeCompressed()
+
+		// Look for a derived key entry and check if it's valid.
+		derivedKeyEntry := bav._getDerivedKeyMappingForOwner(ownerPublicKey, derivedPublicKey)
```

Comment:
This if statement is a little tricky, and I think it's missing the isDeleted case. I think we should do something a little simpler like:

```
if derivedKeyEntry == nil || derivedKeyEntry.isDeleted == true {
return RuleErrorDerivedKeyInvalid
}
```

Reply:
I'm using `OperationType` to determine if the key is authorized or not. I was thinking that having `isDeleted` in the `derivedKeyEntry` might cause confusion since in other txns we're using it for tombstone entries. When we remove a `derivedKeyEntry` from UtxoView in `_disconnectAuthorizeDerivedKey()` it'll revert to the previous entry, or if none exists, it'll set `OperationType` to `OperationValid` and `expirationBlock` to `0` which by default fails the expiration check. This design choice allows us to treat `OperationType` as a one-time switch, where if you de-authorize a key, you can't ever authorize it again. Let me know if you still would like the `isDeleted` field.